packages/video_player/video_player_windows/windows/video_player.cpp

```diff
@@ -0,0 +1,246 @@
+      uri.c_str(), source_resolution_flags, nullptr, &object_type,
+      reinterpret_cast<IUnknown**>(media_source.put_void())));
+  media_engine_wrapper_->Initialize(adapter, media_source.get());
+}
+
+void VideoPlayer::Init(
+    flutter::BinaryMessenger* messenger,
+    std::function<void(int64_t)> texture_frame_available_callback,
+    flutter::TextureRegistrar* texture_registry) {
+  texture_id_ = texture_registry->RegisterTexture(&texture_);
+  texture_frame_available_callback_ = texture_frame_available_callback;
+
+  event_channel_ =
+      std::make_unique<flutter::EventChannel<flutter::EncodableValue>>(
+          messenger,
+          std::string("flutter.io/videoPlayer/videoEvents") +
+              std::to_string(texture_id_),
+          &flutter::StandardMethodCodec::GetInstance());
+
+  event_channel_->SetStreamHandler(
+      std::make_unique<
+          flutter::StreamHandlerFunctions<flutter::EncodableValue>>(
+          [this](const flutter::EncodableValue* arguments,
+                 std::unique_ptr<flutter::EventSink<flutter::EncodableValue>>&&
+                     events)
+              -> std::unique_ptr<
+                  flutter::StreamHandlerError<flutter::EncodableValue>> {
+            event_sink_ = std::move(events);
+            return nullptr;
+          },
+          [this](const flutter::EncodableValue* arguments)
+              -> std::unique_ptr<
+                  flutter::StreamHandlerError<flutter::EncodableValue>> {
+            event_sink_ = nullptr;
+            return nullptr;
+          }));
+}
+
+FlutterDesktopGpuSurfaceDescriptor* VideoPlayer::ObtainDescriptorCallback(
+    size_t width, size_t height) {
+  // Lock buffer mutex to protect texture processing.
+  std::lock_guard<std::mutex> buffer_lock(buffer_mutex_);
+
+  media_engine_wrapper_->UpdateSurfaceDescriptor(
+      static_cast<uint32_t>(width), static_cast<uint32_t>(height),
+      [&]() { texture_frame_available_callback_(texture_id_); }, descriptor_);
+
+  UpdateVideoSize();
+
+  return &descriptor_;
+}
+
+void VideoPlayer::OnMediaInitialized() {
+  // Start playback.
+  media_engine_wrapper_->SeekTo(0);
+  if (!is_initialized_) {
+    is_initialized_ = true;
+    SendInitialized();
+  }
+}
+
+void VideoPlayer::UpdateVideoSize() {
+  auto lock = lock_.lock();
+
+  RECT rect;
+  if (GetWindowRect(window_, &rect)) {
+    int width = rect.right - rect.left;
+    int height = rect.bottom - rect.top;
+    window_size_ = {static_cast<float>(width), static_cast<float>(height)};
+  }
+
+  // If the window has not been initialized yet, use a default size of 640x480.
+  const bool window_initialized =
+      window_size_.Width != 0 && window_size_.Height != 0;
+  float width = window_initialized ? window_size_.Width : 640;
+  float height = window_initialized ? window_size_.Height : 480;
+
+  if (media_engine_wrapper_) {
+    // Call into media engine wrapper on MTA thread to resize the video surface.
+    RunSyncInMTA([&]() {
```

Comment:
Is this necessary? It looks like `MediaEngineWrapper::OnWindowUpdate` also calls `RunSyncInMTA`.

Reply:
Nice catch. Fixed.